plugin/metrics/prometheus/metricsstore/reader.go

```diff
@@ -63,24 +100,149 @@ func NewMetricsReader(logger *zap.Logger, hostPort string, connectTimeout time.D
-	return &metrics.MetricFamily{}, nil
+func (m *MetricsReader) GetCallRates(ctx context.Context, requestParams *metricsstore.CallRateQueryParameters) (*metrics.MetricFamily, error) {
+	metricsParams := metricsQueryParams{
+		BaseQueryParameters: requestParams.BaseQueryParameters,
+		metricName:          callsMetricName,
+		metricDesc:          callsMetricDesc,
+		buildPromQuery: func(p promQueryParams) string {
+			return fmt.Sprintf(
+				// Note: p.spanKindFilter can be ""; trailing commas are okay within a timeseries selection.
+				`sum(rate(calls_total{service_name =~ "%s", %s}[%s])) by (%s)`,
+				p.serviceFilter,
+				p.spanKindFilter,
+				p.rate,
+				p.groupBy,
+			)
+		},
+	}
+	return m.executeQuery(ctx, metricsParams)
 }
 
 // GetErrorRates gets the error rate metrics for the given set of error rate query parameters.
-func (m *MetricsReader) GetErrorRates(ctx context.Context, params *metricsstore.ErrorRateQueryParameters) (*metrics.MetricFamily, error) {
-	// TODO: Implement me
-	return &metrics.MetricFamily{}, nil
+func (m *MetricsReader) GetErrorRates(ctx context.Context, requestParams *metricsstore.ErrorRateQueryParameters) (*metrics.MetricFamily, error) {
+	metricsParams := metricsQueryParams{
+		BaseQueryParameters: requestParams.BaseQueryParameters,
+		metricName:          errorsMetricName,
+		metricDesc:          errorsMetricDesc,
+		buildPromQuery: func(p promQueryParams) string {
+			return fmt.Sprintf(
+				// Note: p.spanKindFilter can be ""; trailing commas are okay within a timeseries selection.
+				`sum(rate(calls_total{service_name =~ "%s", status_code = "STATUS_CODE_ERROR", %s}[%s])) by (%s) / sum(rate(calls_total{service_name =~ "%s", %s}[%s])) by (%s)`,
+				p.serviceFilter, p.spanKindFilter, p.rate, p.groupBy,
+				p.serviceFilter, p.spanKindFilter, p.rate, p.groupBy,
+			)
+		},
+	}
+	return m.executeQuery(ctx, metricsParams)
 }
 
 // GetMinStepDuration gets the minimum step duration (the smallest possible duration between two data points in a time series) supported.
 func (m *MetricsReader) GetMinStepDuration(_ context.Context, _ *metricsstore.MinStepDurationQueryParameters) (time.Duration, error) {
-	return time.Millisecond, nil
+	return minStep, nil
+}
+
+// executeQuery executes a query against a Prometheus-compliant metrics backend.
+func (m *MetricsReader) executeQuery(ctx context.Context, p metricsQueryParams) (*metrics.MetricFamily, error) {
+	if p.GroupByOperation {
+		p.metricName = strings.Replace(p.metricName, "service", "service_operation", 1)
+		p.metricDesc += " & operation"
+	}
+	promQuery := buildPromQuery(p)
+
+	span, ctx := startSpanForQuery(ctx, p.metricName, promQuery)
+	defer span.Finish()
+
+	queryRange := promapi.Range{
+		Start: p.EndTime.Add(-1 * *p.Lookback),
+		End:   *p.EndTime,
+		Step:  *p.Step,
+	}
+
+	m.logger.Debug("Executing Prometheus query", zap.String("query", promQuery), zap.Any("range", queryRange))
+	mv, warnings, err := m.client.QueryRange(ctx, promQuery, queryRange)
```

Comment:
Yup, I've got a `startSpanForQuery` [above](https://github.com/jaegertracing/jaeger/pull/3004/files#diff-86df35ee8e57407ad9099c4dad0cad44d87234486bc9b9d0a38630baf89deae7R174), or were you referring to something else?

Reply:
Oh, you do have it there and I missed it. I think I expected to have the span wrapping the database call, but I think it's fine here.

Reply:
actually, you make a good point, it is in a bit of an odd place, and makes sense to just wrap the database call. I'll clean this up.

Reply:
Another minor comment, sorry about that, but I would either place the span at the beginning of the function followed by `defer span.Finish()`, or I would really wrap just the function call that I'm interested in measuring. Using defer in this situation here means that you are starting to measure time right before going the QueryRange call, but finishing it when the calling function is returning. It doesn't matter much in our case here, though.

```suggestion
	span, ctx := startSpanForQuery(ctx, p.metricName, promQuery)
	mv, warnings, err := m.client.QueryRange(ctx, promQuery, queryRange)
	span.Finish()
```

Reply:
Good point. Then move the startSpan/defer finish to the top of the func?

Reply:
Ah so it would mean also moving it to the original spot, because I also log the `promQuery` which is created [here](https://github.com/jaegertracing/jaeger/pull/3004/files#diff-86df35ee8e57407ad9099c4dad0cad44d87234486bc9b9d0a38630baf89deae7R172) :).

Another option is to extract these few lines of code into another func, whereby we could place the start and finish at the top.

Reply:
And I guess I know see why it was the way you originally set :-)

Reply:
could also place two `span.Finish()` lines, one for the error case and another for the success, but not sure if that's idiomatic.

Reply:
No, keep the original proposal. I think it does make sense now.